File: utils/ql_common.js
'use strict';

const DEFAULT_SEPARATORS = ['\n', '&', '#'];
const ERROR_DETAIL_LIMIT = 1000;
const SENSITIVE_FIELD_PATTERN = /token|cookie|authorization|password|secret|signature|^sign$|api[_-]?key/i;

function splitAccounts(rawValue, separators = DEFAULT_SEPARATORS) {
  if (!rawValue) {
    return [];
  }

  const normalized = rawValue.replace(/\r\n/g, '\n').replace(/\r/g, '\n').trim();
  if (!normalized) {
    return [];
  }

  for (const separator of separators) {
    if (normalized.includes(separator)) {
      return normalized
        .split(separator)
        .map(item => item.trim())
        .filter(Boolean);
    }
  }

  return [normalized];
}

function splitJsonAccounts(rawValue) {
  if (!rawValue) {
    return [];
  }

  const normalized = rawValue.replace(/\r\n/g, '\n').replace(/\r/g, '\n').trim();
  if (!normalized) {
    return [];
  }

  try {
    JSON.parse(normalized);
    return [normalized];
  } catch {
    return splitAccounts(normalized);
  }
}

function maskSecret(value, keep = 4) {
  const text = String(value || '').trim();
  if (text.length <= keep * 2) {
    return '***';
  }
  return `${text.slice(0, keep)}***${text.slice(-keep)}`;
}

function errorMessage(error) {
  const message = error instanceof Error ? error.message : String(error);
  return message.replace(/([?&][^=\s&]+)=([^&\s]+)/g, '$1=***');
}

function safeErrorDetail(data, isJson) {
  const detail = isJson
    ? JSON.stringify(data, (key, value) => (
      key && SENSITIVE_FIELD_PATTERN.test(key) ? '***' : value
    ))
    : data.raw;
  const safeDetail = errorMessage(detail);
  return safeDetail.length > ERROR_DETAIL_LIMIT
    ? `${safeDetail.slice(0, ERROR_DETAIL_LIMIT)}...`
    : safeDetail;
}

function formatResults(results) {
  return results
    .map(result => `${result.ok ? '成功' : '失败'} | 账号${result.index} | ${result.title}: ${result.message}`)
    .join('\n');
}

async function sendNotification(title, content) {
  console.log(`\n${title}\n${content}`);

  let sendNotify;
  try {
    ({ sendNotify } = require('../sendNotify'));
  } catch (error) {
    console.log('未检测到青龙 sendNotify.js，已降级为控制台输出。');
    return;
  }

  try {
    await sendNotify(title, content);
  } catch (error) {
    console.log(`通知发送失败: ${error.message || error}`);
  }
}

async function requestJson(url, options = {}) {
  const timeout = options.timeout || 15000;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeout);

  try {
    const response = await fetch(url, {
      method: options.method || 'GET',
      headers: options.headers,
      body: options.body,
      signal: controller.signal,
    });

    const text = await response.text();
    let data;
    let isJson = true;
    try {
      data = text ? JSON.parse(text) : {};
    } catch (error) {
      data = { raw: text };
      isJson = false;
    }

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${safeErrorDetail(data, isJson)}`);
    }

    return data;
  } finally {
    clearTimeout(timer);
  }
}

module.exports = {
  splitAccounts,
  splitJsonAccounts,
  maskSecret,
  errorMessage,
  formatResults,
  sendNotification,
  requestJson,
};

File: utils/ql_common.py
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from typing import Callable, Iterable, Sequence

DEFAULT_SEPARATORS = ("\n", "&", "#")
QUERY_VALUE_PATTERN = re.compile(r"([?&][^=\s&]+)=([^&\s]+)")


@dataclass
class AccountResult:
    index: int
    ok: bool
    title: str
    message: str


def split_accounts(raw_value: str | None, separators: Sequence[str] = DEFAULT_SEPARATORS) -> list[str]:
    if not raw_value:
        return []

    normalized = raw_value.replace("\r\n", "\n").replace("\r", "\n").strip()
    if not normalized:
        return []

    for separator in separators:
        if separator in normalized:
            return [item.strip() for item in normalized.split(separator) if item.strip()]

    return [normalized]


def mask_secret(value: str, keep: int = 4) -> str:
    value = value.strip()
    if len(value) <= keep * 2:
        return "***"
    return f"{value[:keep]}***{value[-keep:]}"


def error_message(error: Exception) -> str:
    return QUERY_VALUE_PATTERN.sub(lambda match: f"{match.group(1)}=***", str(error))


def cookie_name(cookie: str, account_index: int) -> str:
    for key in ("pt_pin", "un", "BDUSS", "smzdm_id", "uid"):
        match = re.search(rf"(?:^|;\s*){re.escape(key)}=([^;]+)", cookie)
        if match:
            return f"{key}={mask_secret(match.group(1))}"
    return f"账号{account_index}"


def format_results(results: Iterable[AccountResult]) -> str:
    lines = []
    for result in results:
        status = "成功" if result.ok else "失败"
        lines.append(f"{status} | 账号{result.index} | {result.title}: {result.message}")
    return "\n".join(lines)


def send_notify(title: str, content: str) -> None:
    print(f"\n{title}\n{content}")
    try:
        from notify import send
    except ImportError:
        print("未检测到青龙 notify.py，已降级为控制台输出。")
        return

    try:
        send(title, content)
    except Exception as error:
        print(f"通知发送失败: {error}")


def run_accounts(
    script_name: str,
    account_env_name: str,
    runner: Callable[[str, int], AccountResult],
    error_title: Callable[[str, int], str] | None = None,
) -> int:
    accounts = split_accounts(os.getenv(account_env_name))
    if not accounts:
        send_notify(script_name, f"未配置环境变量 {account_env_name}")
        return 1

    results: list[AccountResult] = []
    for account_index, account in enumerate(accounts, start=1):
        try:
            results.append(runner(account, account_index))
        except Exception as error:
            title = error_title(account, account_index) if error_title else f"账号{account_index}"
            results.append(AccountResult(index=account_index, ok=False, title=title, message=error_message(error)))

    send_notify(script_name, format_results(results))
    return 0 if any(result.ok for result in results) else 1
